File: backend/app/services/orchestrator_service.py
from langgraph.graph import StateGraph, END
from agents.state import GraphState
from agents.diagnosis_agent import DiagnosisAgentNode
from agents.intervention_agent import InterventionAgentNode
from agents.planning_agent import PlanningAgentNode
from agents.evaluation_agent import EvaluationAgentNode
from backend.app.schemas.student import (
    StudentSnapshot,
    OrchestrationResponse,
    DiagnosisResult,
    InterventionResult,
    WeeklyScheduleItem,
    EvaluationPlan,
    TraceLogItem,
)
from backend.app.services.knowledge_base import KnowledgeBase

# 1. Instantiate Graph nodes
diagnosis_node = DiagnosisAgentNode()
intervention_node = InterventionAgentNode()
planning_node = PlanningAgentNode()
evaluation_node = EvaluationAgentNode()


def critic_node(state: GraphState) -> dict:
    """A custom critic node to review the planning node outputs before completing."""
    plan = state["weekly_plan"]
    iterations = state.get("iteration_count", 0)

    if len(plan) < 3 and iterations < 2:
        return {
            "critic_feedback": "Plan is too sparse. Add more focus blocks.",
            "iteration_count": iterations + 1,
        }
    return {
        "critic_feedback": None,
    }


def route_triage(state: GraphState) -> str:
    """Conditional routing logic for the planner-critic loop."""
    if state["critic_feedback"] is not None:
        return "replan"
    return "finish"


# 2. Build and Compile the State Machine
workflow = StateGraph(GraphState)

# Add Nodes
workflow.add_node("diagnose", diagnosis_node)
workflow.add_node("intervene", intervention_node)
workflow.add_node("plan", planning_node)
workflow.add_node("evaluate", evaluation_node)
workflow.add_node("criticize", critic_node)

# Wire Nodes
workflow.set_entry_point("diagnose")
workflow.add_edge("diagnose", "intervene")
workflow.add_edge("intervene", "plan")
workflow.add_edge("plan", "evaluate")
workflow.add_edge("evaluate", "criticize")

# Add Loop
workflow.add_conditional_edges(
    "criticize",
    route_triage,
    {
        "replan": "plan",   # Loop back to planning node
        "finish": END,
    },
)

app_graph = workflow.compile()


def _map_to_response(final_state: dict) -> OrchestrationResponse:
    """Map raw GraphState dict to the validated OrchestrationResponse."""
    # Diagnosis maps cleanly
    diagnosis = DiagnosisResult(**final_state["diagnosis"])

    # Intervention needs key remapping: KB uses concept_id, schema uses concept
    raw_intervention = final_state["selected_intervention"]
    intervention = InterventionResult(
        concept=raw_intervention.get("concept") or raw_intervention.get("concept_id", "unknown"),
        error_tag=raw_intervention.get("error_tag") or raw_intervention.get("misconception_id"),
        strategy=raw_intervention.get("strategy", "Targeted review"),
        activities=raw_intervention.get("activities", []),
        why=raw_intervention.get("why", f"Matched misconception '{raw_intervention.get('misconception_id', 'N/A')}'"),
    )

    # Weekly plan
    weekly_plan = [WeeklyScheduleItem(**item) for item in final_state["weekly_plan"]]

    # Evaluation plan
    evaluation_plan = EvaluationPlan(**final_state["evaluation_plan"])

    # Trace
    trace = [TraceLogItem(**item) for item in final_state["trace"]]

    return OrchestrationResponse(
        profile=final_state["profile"],
        diagnosis=diagnosis,
        selected_intervention=intervention,
        weekly_plan=weekly_plan,
        evaluation_plan=evaluation_plan,
        trace=trace,
    )


class OrchestratorService:
    """Coordinates the LangGraph agent pipeline for a single student snapshot."""

    def run(self, snapshot: StudentSnapshot) -> OrchestrationResponse:
        knowledge_base = KnowledgeBase(topic_id=snapshot.profile.subject)

        # 1. Create Initial State
        initial_state: GraphState = {
            "profile": snapshot.profile.model_dump(),
            "attempts": [attempt.model_dump() for attempt in snapshot.attempts],
            "prior_interventions": snapshot.prior_interventions,
            "knowledge_base": {
                "manifest": knowledge_base.manifest,
                "concepts": knowledge_base.concepts,
                "misconceptions": knowledge_base.misconceptions,
                "interventions": knowledge_base.interventions,
                "problems": knowledge_base.problems,
                "evaluation_rules": knowledge_base.evaluation_rules,
            },
            "diagnosis": {},
            "selected_intervention": {},
            "weekly_plan": [],
            "evaluation_plan": {},
            "trace": [],
            "critic_feedback": None,
            "iteration_count": 0,
        }

        # 2. Run LangGraph Engine
        final_state = app_graph.invoke(initial_state)

        # 3. Map to validated response model
        response = _map_to_response(final_state)

        # 4. Persist to database (best-effort, don't break the pipeline)
        self._persist(snapshot, response)

        return response

    @staticmethod
    def _persist(snapshot: StudentSnapshot, response: OrchestrationResponse) -> None:
        """Best-effort persistence of orchestration results to SQLite."""
        try:
            from backend.app.services.database import (
                upsert_student,
                record_attempt,
                record_intervention,
            )

            # Save student profile
            upsert_student(
                student_id=snapshot.profile.student_id,
                subject=snapshot.profile.subject,
                available_hours=snapshot.profile.available_hours_per_week,
            )

            # Save each attempt
            for attempt in snapshot.attempts:
                record_attempt(
                    student_id=snapshot.profile.student_id,
                    problem_id=attempt.problem_id,
                    concept=attempt.concept,
                    correct=attempt.correct,
                    error_tags=attempt.error_tags,
                    time_seconds=attempt.time_seconds,
                    hints_used=attempt.hints_used,
                    retries=attempt.retries,
                )

            # Save the intervention
            record_intervention(
                student_id=snapshot.profile.student_id,
                concept=response.selected_intervention.concept,
                strategy=response.selected_intervention.strategy,
            )

        except Exception:
            import logging
            logging.getLogger(__name__).warning(
                "Failed to persist orchestration results to database.", exc_info=True,
            )

File: backend-node/src/services/orchestratorService.js
import { KnowledgeBase } from "./knowledgeBase.js";
import { diagnose } from "../agents/diagnosisAgent.js";
import { intervene } from "../agents/interventionAgent.js";
import { plan } from "../agents/planningAgent.js";
import { evaluate } from "../agents/evaluationAgent.js";
import { criticize } from "../agents/criticAgent.js";
import {
  upsertStudent,
  recordAttempt,
  recordIntervention,
  updateMastery,
} from "./database.js";
import { upsertLearnerState, recordAttemptEvent } from "./mongoService.js";
import { buildQuestionGenerationContext } from "./retrievalService.js";

function _mapToResponse(finalState) {
  const diagnosis = finalState.diagnosis;
  const rawIntervention = finalState.selected_intervention;
  
  const intervention = {
    concept: rawIntervention.concept || rawIntervention.concept_id || "unknown",
    error_tag: rawIntervention.error_tag || rawIntervention.misconception_id || null,
    strategy: rawIntervention.strategy || "Targeted review",
    activities: rawIntervention.activities || [],
    why: rawIntervention.why || `Matched misconception '${rawIntervention.misconception_id || "N/A"}'`,
    estimated_sessions: rawIntervention.estimated_sessions || 2,
  };

  return {
    profile: finalState.profile,
    diagnosis,
    selected_intervention: intervention,
    weekly_plan: finalState.weekly_plan || [],
    evaluation_plan: finalState.evaluation_plan || {},
    trace: finalState.trace || [],
    critic_iterations: finalState.iteration_count || 0,
  };
}

export class OrchestratorService {
  /**
   * Coordinates the agent pipeline for a single student snapshot.
   */
  async run(snapshot) {
    if (!snapshot.attempts || snapshot.attempts.length === 0) {
      throw {
        status: 409,
        detail: "A fresh learner needs an initial assessment before the planner can diagnose anything.",
        requires_initial_assessment: true,
      };
    }

    const kb = new KnowledgeBase(snapshot.profile.subject);
    const fallbackConcepts = [...new Set((snapshot.attempts || []).map((a) => a.concept).filter(Boolean))];
    const retrievalContext = await buildQuestionGenerationContext(
      snapshot.profile.student_id,
      fallbackConcepts
    );

    // 1. Create Initial State
    let state = {
      profile: snapshot.profile,
      attempts: snapshot.attempts || [],
      prior_interventions: snapshot.prior_interventions || [],
      knowledge_base: {
        manifest: kb.manifest,
        concepts: kb.concepts,
        misconceptions: kb.misconceptions,
        interventions: kb.interventions,
        problems: kb.problems,
        evaluation_rules: kb.evaluationRules,
      },
      diagnosis: {},
      selected_intervention: {},
      weekly_plan: [],
      evaluation_plan: {},
      retrieval_context: retrievalContext,
      trace: [],
      critic_feedback: null,
      iteration_count: 0,
    };

    // 2. Run Engine (Replacing LangGraph with a standard loop)
    state = await diagnose(state);
    state = await intervene(state);

    while (true) {
      state = await plan(state);
      state = await evaluate(state);
      state = await criticize(state);
      
      if (state.critic_feedback === null) {
        break; // Plan approved or iteration limit reached
      }
      // Replan branch: loops back to plan() since critic_feedback is set
    }

    // 3. Map to validated response model
    const response = _mapToResponse(state);
    response.retrieval_context = retrievalContext;

    // 4. Persist to database (best-effort)
    await this._persist(snapshot, response);

    return response;
  }

  async _persist(snapshot, response) {
    try {
      const { profile, attempts = [] } = snapshot;
      
      // Save student profile
      upsertStudent(
        profile.student_id,
        profile.subject,
        profile.available_hours_per_week || 6
      );

      // Save each attempt
      for (const attempt of attempts) {
        recordAttempt({
          studentId: profile.student_id,
          problemId: attempt.problem_id,
          concept: attempt.concept,
          correct: attempt.correct,
          errorTags: attempt.error_tags,
          timeSeconds: attempt.time_seconds,
          hintsUsed: attempt.hints_used,
          retries: attempt.retries,
        });
      }

      const mastery = profile.confidence_by_concept || {};
      for (const [concept, confidence] of Object.entries(mastery)) {
        updateMastery(profile.student_id, concept, confidence);
      }
      const weakConcepts = Object.entries(mastery)
        .sort((a, b) => a[1] - b[1])
        .slice(0, 5)
        .map(([concept]) => concept);
      await upsertLearnerState({
        studentId: profile.student_id,
        profile,
        confidenceByConcept: mastery,
        weakConcepts,
      });

      // Save the intervention
      if (response.selected_intervention) {
        recordIntervention({
          studentId: profile.student_id,
          concept: response.selected_intervention.concept,
          strategy: response.selected_intervention.strategy,
        });
      }

      for (const attempt of attempts) {
        await recordAttemptEvent({
          student_id: profile.student_id,
          problem_id: attempt.problem_id,
          concept: attempt.concept,
          correct: Boolean(attempt.correct),
          error_tags: attempt.error_tags || [],
          time_seconds: attempt.time_seconds ?? null,
        });
      }
    } catch (error) {
      console.warn("Failed to persist orchestration results to database:", error);
    }
  }
}
